source/opt/folding_rules.h

```diff
@@ -0,0 +1,53 @@
+// Copyright (c) 2018 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef LIBSPIRV_UTIL_FOLDING_RULES_H_
+#define LIBSPIRV_UTIL_FOLDING_RULES_H_
+
+#include "constants.h"
+#include "def_use_manager.h"
+#include "ir_builder.h"
+#include "ir_context.h"
+
+#include <cstdint>
+#include <vector>
+
+namespace spvtools {
+namespace opt {
+
+using FoldingRule =
```

Comment:
This needs an explanation, most of all.  What is the bool for?  What are the arguments"?

Reply:
Yes, I wanted to add a detailed description on how to add a new folding rule.